:sparkles: Replace `Timeout` with `RolloutConfig`

dhaiducek opened the pull request:
<!--
:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization
-->

## Summary

Implement `RolloutConfig`, which provides more fine-tuned configuration for users than the current `Timeout`.

## Related issue(s)

- https://github.com/open-cluster-management-io/enhancements/pull/100

qiujian16 commented:
@dhaiducek since we are about to release 0.12.0, would like to know the progress of this PR. Is it ready for review, do you plan to include it in 0.12.0 or it can be moved to next release?

dhaiducek commented:
@qiujian16 It was waiting on #276, but now that it's merged I'll continue work here. I'll work on it this week, but I think it's also fine if `v0.12.0` is released without these library updates so that this PR doesn't block the release.

dhaiducek commented:
Status update: I wrestled with the CI a bit because I'm on an M1 MacBook and some of the binaries aren't actually built for arm64 (those changes are in this PR). I've got current code working, migrating `Timeout` --> `ProgressDeadline`, but need to add logic (and tests) for the other fields that are in the new struct alongside it.

dhaiducek committed:
Replace Timeout with RolloutConfig

Uses `MinSuccessTime`, `ProgressDeadline`, and `MaxFailures` for a more
fine-tuned configuration.
Also passes `Timeout` over to `ProgressDeadline` and marks it as
deprecated.

Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

dhaiducek commented:
/cc @serngawy @haoqing0110 

Sorry for the long delay here--between wrestling with the CI and getting pulled away for other tasks, it's taken longer to update than I thought. There might be some more tests required here, but I think the logic is largely (if not entirely) there.

Some notes:
- I added logic to pass `Timeout` over to `ProgressDeadline` and mark `Timeout` as deprecated. I debated having `MaxFailures` set to `100%` so the behavior wouldn't change there, but I think it better to leave it and make the migration obvious.
- I made a decision to handle `MaxFailures` over the total number of clusters for `Progressive` and per group for `ProgressivePerGroup`, partially because `Progressive` clusters are sorted alphabetically, so having a cluster injected at the beginning or middle could complicate things.

serngawy committed:
Add more unit test for Rollout ProgressivePerGroup case

Signed-off-by: melserngawy <melserng@redhat.com>

serngawy commented:
Sounds good to me in general, I added more unit test to cover Rollout ProgressivePerGroup  case.

dhaiducek committed:
Fix ProgressivePerGroup handling

- MandatoryDecisionGroups should tolerate no failures
- The logic for ProgressivePerGroup was incorrect

Signed-off-by: Dale Haiducek <19750917+dhaiducek@users.noreply.github.com>

dhaiducek commented:
After discussion with @serngawy I've updated the logic so that `MandatoryDecisionGroups` tolerate no failures for any progressive rollout type.

haoqing0110 commented on cluster/v1alpha1/helpers.go:
```diff
@@ -181,13 +205,21 @@ func (r *RolloutHandler[T]) getProgressivePerGroupClusters(rolloutStrategy Rollo
 	}
 
+	// Check format of MaxFailures--this value will be re-parsed and used in progressivePerGroup()
+	_, err = calculateRolloutSize(maxFailures, 100, 0)
```
Can understand the intention here. Is that possible to split the calculateRolloutSize into a parse function and a calculate function, so that don't need to hardcode a 100 which looks strange.

dhaiducek replied:
Ah, good point--I'll update it.

haoqing0110 commented:
LGTM, seems the `ProgressDeadline` is just a rename of `Timeout`, I'm fine to leave the `Timeout` and mark it as deprecated.

qiujian16 commented:
/approve

Since manifestworkReplicaSet and ClusterManagerAddon install strategy  is in alpha stage. Let's remove timeout in the next release (ie 0.14.0)

dhaiducek commented:
> Since manifestworkReplicaSet and ClusterManagerAddon install strategy is in alpha stage. Let's remove timeout in the next release (ie 0.14.0)

Sounds good, thanks @qiujian16!

@serngawy @haoqing0110 I'd appreciate if you'd do another review here when you have a moment! 😄

serngawy commented:
> > Since manifestworkReplicaSet and ClusterManagerAddon install strategy is in alpha stage. Let's remove timeout in the next release (ie 0.14.0)
> 
> Sounds good, thanks @qiujian16!
> 
> @serngawy @haoqing0110 I'd appreciate if you'd do another review here when you have a moment! 😄

I gave it another review, sounds good to me @dhaiducek appreciate the good work :) 

ops, didn't consider my approve will merge it @haoqing0110
